lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1883,25 +1890,41 @@ class ImplicitSelfUsageChecker : public BaseDiagnosticWalker {
       }
     }
 
-    // If the self decl comes from a conditional statement, validate
-    // that it is an allowed `guard let self` or `if let self` condition.
+    // If the self decl refers to a `guard let self` / `if let self` conditon,
+    // we need to validate it.
+    auto selfDeclDefinedInConditionalStmt = false;
+    auto isInvalidSelfDeclRebinding = false;
+    if (auto condStmt = parentConditionalStmt(selfDecl)) {
+      selfDeclDefinedInConditionalStmt = true;
+      isInvalidSelfDeclRebinding = !hasValidSelfRebinding(condStmt, ctx);
+    }
+
+    // If the self decl refers to an invalid unwrapping conditon like
+    // `guard let self = somethingOtherThanSelf`, then implicit self is always
+    // disallowed.
     //  - Even if this closure doesn't have a `weak self` capture, it could
     //    be a closure nested in some parent closure with a `weak self`
     //    capture, so we should always validate the conditional statement
     //    that defines self if present.
-    if (validateSelfRebindings) {
-      if (auto conditionalStmt = parentConditionalStmt(selfDecl)) {
-        if (!hasValidSelfRebinding(conditionalStmt, ctx)) {
-          return false;
-        }
-      }
+    if (isInvalidSelfDeclRebinding) {
+      return false;
     }
 
-    // If this closure has a `weak self` capture, require that the
-    // closure unwraps self. If not, implicit self is not allowed
-    // in this closure or in any nested closure.
+    // Within a closure with a `[weak self]` capture, implicit self is only
+    // allowed if self has been unwrapped in a previous conditional statement.
+    //  - When validating implicit self usage in a nested closure, it's not
+    //    necessary for self to be unwrapped in this parent closure. If self
+    //    isn't unwrapped correctly in the nested closure, we would have
+    //    already noticed when validating that closure.
     if (closureHasWeakSelfCapture(inClosure) &&
-        !hasValidSelfRebinding(parentConditionalStmt(selfDecl), ctx)) {
+        !selfDeclDefinedInConditionalStmt && !isValidatingParentClosures) {
+      return false;
+    }
+
+    // If the self decl refers to a weak capture in a parent closure,
+    // then implicit self is not allowed.
+    if (selfDecl->getInterfaceType()->is<WeakStorageType>() &&
+        !closureHasWeakSelfCapture(inClosure)) {
```

Comment:
I think these checks can be consolidated, the first is indirectly checking whether we're directly referencing a `weak self` capture in the current closure, the second is checking that for a parent closure. I think this could be simplified to:

```cxx
if (selfDecl->getInterfaceType()->is<WeakStorageType>() && !isValidatingParentClosures)
```

That way _any_ implicit reference to a `weak self` is rejected. What do you think?

Reply:
Good point, the other check is unnecessary now. Thanks for the suggestion!